sklearn/impute/tests/test_impute.py

```diff
@@ -1786,3 +1786,57 @@ def test_simple_imputer_constant_fill_value_casting():
         )
         X_trans = imputer.fit_transform(X_float32)
         assert X_trans.dtype == X_float32.dtype
+
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_no_empty_features(strategy):
+    """Check the behavior of the iterative imputer.
+    When the input matrix contains no empty features, it should return
+    the same result regardless of whether `keep_empty_features`.
+    """
+    rng = np.random.RandomState(0)
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
+    X[X == 0] = np.nan
+
+    imputer_false = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=False
+    )
+    X_imputed_false = imputer_false.fit_transform(X)
+
+    imputer_true = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=True
+    )
+    X_imputed_true = imputer_true.fit_transform(X)
+
+    assert_array_equal(X_imputed_false, X_imputed_true)
+
+
+@pytest.mark.parametrize("strategy", ["mean", "median", "most_frequent", "constant"])
+def test_iterative_imputer_with_empty_features(strategy):
+    """Check the behavior of the iterative imputer.
+    When the input matrix contains empty features, it should return
+    the same values for the non-empty features regardless of
+    `keep_empty_features`. Additionally, the empty features
+    should become 0s when `keep_empty_features` is set to True.
+    """
+    rng = np.random.RandomState(0)
+    X = _sparse_random_matrix(10, 10, density=0.95, random_state=rng).toarray()
+    X[X == 0] = np.nan
+    X[:, 0] = np.nan
+
+    imputer_false = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=False
+    )
+    X_imputed_false = imputer_false.fit_transform(X)
+
+    imputer_true = IterativeImputer(
+        initial_strategy=strategy, fill_value=1, keep_empty_features=True
```

Comment:
I would still use `fill_value=0` to not have a special casing for `constant` when checking the value.

Reply:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/31cf853d7cf4ed09daa1555cc1d5b3196ee1b681